subsystems/DiscordRPC/src/main/java/org/terasology/engine/subsystem/discordrpc/DiscordRPCThread.java

```diff
@@ -0,0 +1,284 @@
+
+public final class DiscordRPCThread implements IPCListener, Runnable {
+    private static final Logger logger = LoggerFactory.getLogger(DiscordRPCThread.class);
+    private static final long DISCORD_APP_CLIENT_ID = 515274721080639504L;
+    private static final String DISCORD_APP_DEFAULT_IMAGE = "ss_6";
+    private static final int MAX_RECONNECT_TRIES = 5;
+
+    private final Thread thread;
+    private final IPCClient ipcClient;
+    private final DiscordRPCBuffer buffer;
+    private RichPresence lastRichPresence;
+
+    private int tries;
+
+    private boolean enabled;
+    private boolean waiting;
+    private boolean connectedBefore;
+    private boolean connected;
+    private boolean autoReconnect;
+
+    public DiscordRPCThread() {
+        thread = new Thread(this);
+        thread.setName("DISCORD-RPC-THREAD");
+
+        ipcClient = new IPCClient(DISCORD_APP_CLIENT_ID);
+        ipcClient.setListener(this);
+
+        buffer = new DiscordRPCBuffer();
+
+        lastRichPresence = null;
+
+        tries = 0;
+
+        enabled = false;
+        waiting = false;
+        connectedBefore = false;
+        connected = false;
+        autoReconnect = false;
+    }
+
+    public void start() {
+        thread.start();
+    }
+
+    public synchronized void stop() {
+        synchronized (thread) {
+            thread.interrupt();
+        }
+    }
+
+    public synchronized void discover() {
+        if (enabled && connected) {
+            return;
+        }
+
+        reset(true);
+
+        connectedBefore = true;
+    }
+
+    public synchronized void enable() {
+        if (enabled) {
+            return;
+        }
+
+        enabled = true;
+        autoReconnect = true;
+
+        if (waiting && thread.isAlive()) {
+            synchronized (thread) {
+                thread.notify();
+            }
+        }
+    }
+
+    public synchronized void disable() {
+        disable(false);
+    }
+
+    public synchronized void disable(boolean keepConnectionAlive) {
+        if (!enabled) {
+            return;
+        }
+
+        enabled = false;
+
+        reset(!keepConnectionAlive);
+        if (!keepConnectionAlive) {
+            autoReconnect = false;
+        }
+
+        if (waiting && thread.isAlive()) {
+            synchronized (thread) {
+                thread.notify();
+            }
+        }
+    }
+
+    @Override
+    public void onReady(IPCClient ignored) {
+        if (connectedBefore) {
+            logger.info("Re-connected to Discord RPC!");
+        } else {
+            logger.info("Connected to Discord RPC!");
+        }
+
+        connectedBefore = true;
+        connected = true;
+    }
+
+    @Override
+    public void onDisconnect(IPCClient client, Throwable t) {
+        connected = false;
+        logger.info("Discord RPC lost connection: Disconnected!");
+    }
+
+    @Override
+    public void run() {
+        while (true) {
+            logger.info("Waiting for auto-connecting...");
```

Comment:
How much will this actually log on the INFO level? I'm cautious when I see `logger.info` in combination with loops or frequently called methods as this might spam the log.

Reply:
The purpose of the loop is to keep the thread alive. It's rare for this auto-connect log to appear.

It will appear by one of these conditions below:
- Once on the start
- Every time the player disables the rich presence from the player settings (if the player doesn't have the Discord client)
- When the thread tries to find the Discord client but still couldn't connect to it.

Reply:
Likewise a bit concerned here, it doesn't seem overly far fetched that a connectivity issue could appear then suddenly the log is being filled by junk due to a decidedly optional extra that you could play without perfectly fine otherwise :-)

In other words: if the Discord integration breaks or is in a bad state it shouldn't be _too_ loud about it - ideally just log once per "incident" however we define that

Reply:
So what happens if the discord servers are down, or the player just does not have an active internet connection?

Reply:
The Discord client hosts a local IPC server for the applications to connect it.
When the application connects it exchanges information with the server to identify the application.
And then the application can send RPC information every x amount of time to update it.
The Discord client decides which application RPC will be shown depends on different factors.

If something happens in the Discord servers the only affected side is going to be the Discord client.
The Discord Client runs the IPC server when the client starts. And it doesn't stop until the client is being shutdown.
And until that happens it's going to stay listening to application RPCs whether the Discord server is down or not.